docs: Add "Introduction to Helm" conceptual overview page (v3 and v4) (#2171)
helm/helm-www: docs/intro/introduction.md

This lets you treat an application the way a system package manager such as
Homebrew, apt, or yum treats software on an operating system.

## What can Helm do?

You can use Helm to:

- **Install and manage applications.** Deploy off-the-shelf applications, such
  as databases, monitoring stacks, and ingress controllers, from a chart
  instead of assembling manifests yourself.
- **Package and share your own applications.** Bundle your Kubernetes resources
  into a chart, version it, and distribute it to your team or the wider
  community.
- **Configure applications per environment.** Supply different values to the
  same chart to deploy to development, staging, and production without
  duplicating manifests.
- **Upgrade and roll back safely.** Move a release to a new version, and return
  to a previous revision when an upgrade does not go as planned.
- **Manage dependencies.** Declare the other charts your application needs, and
  let Helm install them together.

## Who is Helm For?

A Helm user often performs one of several roles.
One person can perform more than one of these roles, and how the roles map to
people varies between organizations.

> These roles are drawn from the Helm [User Profiles](/community/user-profiles).

Reviewer: ```suggestion
For more information, see [User Profiles](/community/user-profiles) in the Helm Community documentation.
```

### Chart

A _chart_ is a Helm package.
It contains all of the resource definitions needed to run an application, tool,
or service inside a Kubernetes cluster.
Think of it as the Kubernetes equivalent of a Homebrew formula, an apt `dpkg`,
or a yum RPM file.
To learn how to build one, see the [Charts guide](/topics/charts.mdx).

### Repository

A _repository_ is the place where charts are collected and shared.
It works like Perl's [CPAN archive](https://www.cpan.org) or the
[Fedora Package Database](https://src.fedoraproject.org/), but for Kubernetes
packages.
You can find publicly available charts on [Artifact Hub](https://artifacthub.io),
which lists charts from many repositories.
For more information about hosting charts, see the
[Chart Repository guide](/topics/chart_repository.md).

### Release

A _release_ is an instance of a chart running in a Kubernetes cluster.
You can install one chart many times into the same cluster, and each
installation creates a new release with its own release name.
For example, if you want two databases running in your cluster, you can install
a MySQL chart twice, and each installation is tracked as a separate release.

> When Helm creates a release, it merges a _configuration_ into the chart:

Reviewer: ```suggestion
To create a release, Helm merges a packaged chart with configuration information.
```

> the set of values, typically from a `values.yaml` file.

Reviewer: ```suggestion
The configuration is a set of values, typically from a `values.yaml` file.
```

> Supplying different values to the same chart produces different releases.

Reviewer: ```suggestion
```

---
title: Introduction to Helm
description: A conceptual overview of what Helm is, why you use it, who it is for, and how it works.
sidebar_position: 1
---

# Introduction to Helm

Helm is the package manager for Kubernetes.
It helps you define, install, and upgrade applications on a Kubernetes cluster,
from a single container to an application with many interdependent parts.

Deploying an application to Kubernetes means writing and maintaining many
manifests: Deployments, Services, ConfigMaps, and more.
Managing those files by hand across environments and versions is repetitive and
error prone.
Helm packages these related manifests into a single unit called a _chart_,
which you can version, share, install, and roll back as one release.
This lets you treat an application the way a system package manager such as
Homebrew, apt, or yum treats software on an operating system.

To read more about the concepts introduced on this page, see the

> [Glossary](/glossary/index.mdx).

Reviewer: @promptless-for-oss delete

> ## Common Use Cases

Reviewer: @promptless-for-oss let's change the heading to "What can Helm do?"

Common use cases would imply really specific examples, but these are more of a high level overview of what Helm is used for

You can use Helm to:

- **Install and manage applications.** Deploy off-the-shelf applications, such
  as databases, monitoring stacks, and ingress controllers, from a chart
  instead of assembling manifests yourself.
- **Package and share your own applications.** Bundle your Kubernetes resources
  into a chart, version it, and distribute it to your team or the wider
  community.
- **Configure applications per environment.** Supply different values to the
  same chart to deploy to development, staging, and production without
  duplicating manifests.
- **Upgrade and roll back safely.** Move a release to a new version, and return
  to a previous revision when an upgrade does not go as planned.
- **Manage dependencies.** Declare the other charts your application needs, and
  let Helm install them together.

> ## Who Helm Is For

Author: The "Who Helm Is For" section is adapted from the Helm `user-profiles.md` community page, using the Application operator, Application distributor, Application developer, and Supporting tool developer profiles plus the out-of-scope Cluster operator note. The Helm developer profile was intentionally omitted per the request.

Source: https://github.com/helm/helm-www/blob/main/community/user-profiles.md

> ## Who Helm Is For

Reviewer: @promptless-for-oss let's change to Who is Helm For? to match the rewritten What can Helm do? heading

A Helm user often performs one of several roles.
One person can perform more than one of these roles, and how the roles map to
people varies between organizations.
These roles are drawn from the Helm [User Profiles](/community/user-profiles).

- **Application operator.** You take an application and run it inside a
  Kubernetes cluster, such as operating WordPress and its MySQL database.
  This role differs from a cluster operator, who runs the cluster itself.
- **Application distributor.** You package an application so that someone else
  can operate it, as the maintainers of community charts do.
- **Application developer.** You write the software for an application, and are
  typically not concerned with where it runs.
- **Supporting tool developer.** You build tools that work alongside Helm, such
  as a linter or a Helm plugin.

Helm focuses on the application running in the cluster rather than the cluster
itself.
Standing up and operating a Kubernetes cluster, including its control plane and
nodes, is the work of a cluster operator and falls outside Helm's scope.

> ## Architecture

Author: The "Architecture" section consolidates content from the retired `docs/topics/architecture.md` page: the list of what Helm can do, the chart/config/release inputs, the Helm client vs. Helm library split, and the Go/REST+JSON/Secrets implementation details. That page is deleted in this change and redirected to this overview.

Source: https://github.com/helm/helm-www/blob/main/docs/topics/architecture.md

Helm is a command-line tool that runs on your local machine and talks to the
[Kubernetes API server](https://kubernetes.io/docs/concepts/overview/kubernetes-api/).
It manages Kubernetes packages called charts, and it can:

- Create new charts from scratch.
- Package charts into chart archive (`.tgz`) files.
- Interact with chart repositories where charts are stored.
- Install charts into a Kubernetes cluster, and uninstall them.

> - Manage the release cycle of charts that have been installed with Helm.

Reviewer: @promptless-for-oss this section needs to be revisited. I see that it pulls information from https://helm.sh/docs/topics/architecture#the-purpose-of-helm, which is more about describing the purpose of Helm and is not strictly related to "Architecture"

In order to reduce unnecessary redundancies on this page, my suggestion is to remove this portion from this Architecture section. It can probably just be deleted, but it's worth considering if some of this information can/should be incorporated into the existing Common Use Cases or Components sections

Helm builds a release from two inputs: a _chart_ and a _configuration_.
A chart bundles the information needed to create an instance of an application.
A configuration holds the values merged into the chart to produce a releasable
object.
A _release_ is a running instance of a chart combined with a specific

> configuration.

Reviewer: @promptless-for-oss the way this information is framed conflicts with the idea of charts, releases, and repositories being the three key components (which is stated in the Key Components section below). Instead, this section posits that the main components of Helm are actually a chart and a configuration (or even maybe a chart, a configuration, and a release? It's not very clear).

I notice that the only new information presented here is that a configuration holds values (presumably from the values.yaml file?), and that Helm uses these values to create a release.

My suggestion is to heavily condense this information so that it only covers the new information about how Helm uses configurations/values when creating a release. Then,  relocate it to the Key Components > Releases section below.

Helm is built in two distinct parts:

- **The Helm client** is the command-line tool for end users.
  It handles local chart development, manages repositories, manages releases,
  and sends charts to the Helm library to be installed, upgraded, or
  uninstalled.
- **The Helm library** provides the logic that carries out Helm operations.

>   It combines a chart and a configuration into a release, and it installs,

Reviewer: @promptless-for-oss related to https://github.com/helm/helm-www/pull/2171/changes#r3538636724, I'm not sure it's necessary to include detail about "...combines a chart and a configuration..." here. It just adds confusion. The important information when it comes to the Helm library for this section is just that it's the part that creates releases. Users can then read about releases if they want more information about what they are.

  upgrades, and uninstalls charts by interacting with the Kubernetes API server.
  Because the library is standalone, other clients can reuse the same logic.

The Helm client and library are written in the [Go](https://go.dev) programming
language, and the library uses the Kubernetes client library to communicate with
Kubernetes over REST and JSON.
Helm stores release information in Kubernetes
[Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) inside the
cluster, so it does not need its own database.
Configuration files are written in [YAML](https://yaml.org) where possible.

> ## Key Components

Reviewer: @promptless-for-oss let's move Key Components just above the Architecture section instead. The Architecture section assumes people know what charts and releases are.

Three components describe how Helm works:
charts, repositories, and releases.

> ### Chart

Author: The Chart, Repository, and Release definitions are drawn from the "Three Big Concepts" section of `using_helm.mdx`, which this change removes in favor of this overview (replacing it with a cross-reference to this page).

Source: https://github.com/helm/helm-www/blob/main/docs/intro/using_helm.mdx

or service inside a Kubernetes cluster.
Think of it as the Kubernetes equivalent of a Homebrew formula, an apt `dpkg`,
or a yum RPM file.
To learn how to build one, see the [Charts guide](/topics/charts.mdx).

### Repository

A _repository_ is the place where charts are collected and shared.
It works like Perl's [CPAN archive](https://www.cpan.org) or the
[Fedora Package Database](https://src.fedoraproject.org/), but for Kubernetes
packages.
You can find publicly available charts on [Artifact Hub](https://artifacthub.io),
which lists charts from many repositories.
For more information about hosting charts, see the
[Chart Repository guide](/topics/chart_repository.md).

### Release

A _release_ is an instance of a chart running in a Kubernetes cluster.
You can install one chart many times into the same cluster, and each
installation creates a new release with its own release name.
For example, if you want two databases running in your cluster, you can install
a MySQL chart twice, and each installation is tracked as a separate release.

With these components in mind, you can describe Helm this way:
Helm installs charts into Kubernetes, creating a new release for each

> installation, and you find new charts by searching Helm chart repositories.

Reviewer: @promptless-for-oss as is, this content appears below the Release heading, which makes it seem like it only applies to releases. Instead, let's do this:
- Move it just after "Three components describe how Helm works: charts, repositories, and releases."
- Delete "With these components in mind, you can describe Helm this way:"

The full intro paragraph would then be "Three components describe how Helm works: charts, repositories, and releases. Helm installs charts into Kubernetes, creating a new release for each installation, and you find new charts by searching Helm chart repositories."
